### PR TITLE
OpenShift container image extension

### PR DESCRIPTION
**Pull request by iocanel**
This is work in progress ...

The pull request introduces the docker strategy to openshift builds.

The idea is to introduce `container-image-openshift` extension that will support two strategies
- docker 
- s2i binary build

At the moment everything is implemented inside s2i extension. I intend to split in follow up pull requests, but since its the first working draft, I thought I'd share with the rest of the team.

### CODE REVIEW

**Comment by iocanel** [comment]
@geoand @maxandersen FYI

**Comment by geoand** [comment]
Great! I'll take a look in the morning

File: integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/OpenshiftWithDockerBuildStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Test
+    public void assertGeneratedResources() throws IOException {
+        Path kubernetesDir = prodModeTestResults.getBuildDir().resolve("kubernetes");
+
+        assertThat(kubernetesDir).isDirectoryContaining(p -> p.getFileName().endsWith("openshift.json"))
+                .isDirectoryContaining(p -> p.getFileName().endsWith("openshift.yml"));
+        List<HasMetadata> openshiftList = DeserializationUtil.deserializeAsList(kubernetesDir.resolve("openshift.yml"));
+
+        assertThat(openshiftList).filteredOn(h -> "BuildConfig".equals(h.getKind())).hasOnlyOneElementSatisfying(h -> {
```
Comment (geoand) [inline integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/OpenshiftWithDockerBuildStrategyTest.java:45]: Super minor nitpick: Best use `single().satisfies(h -> {...})` as `hasOnlyOneElementSatisfying` has been deprecated

**Comment by geoand** [comment]
I did a really quick pass and I like the general idea a lot. 
I'll do a more deep dive tomorrow

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java [hunk 2]
```diff
@@ -65,12 +83,36 @@ public CapabilityBuildItem capability() {
...
+        }
+    }
+
+    @BuildStep(onlyIf = { S2iBuild.class, NativeBuild.class })
+    public void s2iPrepareNativeDockerBuild(S2iConfig s2iConfig,
+            OutputTargetBuildItem out,
+            BuildProducer<DecoratorBuildItem> decorator) {
+        if (s2iConfig.buildStrategy == BuildStrategy.docker) {
+            decorator.produce(new DecoratorBuildItem(new ApplyDockerfileToBuildConfigDecorator(null,
+                    findMainSourcesRoot(out.getOutputDirectory()).getValue().resolve(s2iConfig.nativeDockerfile))));
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java:103]: We should probably move this some utility common to all container-image extensions - It doesn't need to be now of course

**Comment by geoand** [comment]
What would also be very useful in reviewing, is if you pasted an example output of the BuildConfig.

Also what problems have you encountered so far? Where does the current PR fall short?

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java [hunk 3]
```diff
@@ -65,12 +83,36 @@ public CapabilityBuildItem capability() {
...
+        decorator.produce(new DecoratorBuildItem(new RemoveEnvVarDecorator(null, "JAVA_APP_JAR")));
+    }
+
     @BuildStep(onlyIf = { IsNormal.class, S2iBuild.class }, onlyIfNot = NativeBuild.class)
     public void s2iRequirementsJvm(S2iConfig s2iConfig,
             CurateOutcomeBuildItem curateOutcomeBuildItem,
             OutputTargetBuildItem out,
             PackageConfig packageConfig,
             JarBuildItem jarBuildItem,
+            BuildProducer<DecoratorBuildItem> decorator,
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java:115]: This seems to be unused

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jsoup.helper.StringUtil;
+
+import io.dekorate.DekorateException;
+import io.dekorate.deps.commons.compress.archivers.tar.TarArchiveEntry;
+import io.dekorate.deps.commons.compress.archivers.tar.TarArchiveOutputStream;
+
+public class PackageUtil {
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:26]: Can we get some comments in this?

Mentioning it because this looks like where most of the problems could arise - the rest of the changes seem pretty straighforward, but this one is a mouth full :)

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    public static TarArchiveOutputStream buildTarStream(File outputPath) throws IOException {
+        FileOutputStream fout = new FileOutputStream(outputPath);
+        BufferedOutputStream bout = new BufferedOutputStream(fout);
+        //BZip2CompressorOutputStream bzout = new BZip2CompressorOutputStream(bout);
+        TarArchiveOutputStream stream = new TarArchiveOutputStream(bout);
+        stream.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
+        return stream;
+    }
+
+    public static void tar(Path inputPath, Path outputPath) throws IOException {
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:148]: This seems to be unused

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        if (!Files.exists(inputPath)) {
+            throw new FileNotFoundException("File not found " + inputPath);
+        }
+
+        try (TarArchiveOutputStream tarArchiveOutputStream = buildTarStream(outputPath.toFile())) {
+            if (!Files.isDirectory(inputPath)) {
+                TarArchiveEntry tarEntry = new TarArchiveEntry(inputPath.toFile().getName());
+                if (inputPath.toFile().canExecute()) {
+                    tarEntry.setMode(tarEntry.getMode() | 0755);
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:157]: Are you sure using `0755` like this is correct?

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+                        }
+                        final Path relativePath = root.relativize(file);
+                        final boolean hasBasePath = !StringUtil.isBlank(base);
+                        final TarArchiveEntry entry = hasBasePath
+                                ? new TarArchiveEntry(base + File.separator + file.toFile())
+                                : new TarArchiveEntry(file.toFile());
+
+                        entry.setName(hasBasePath ? base + File.separator + relativePath.toString() : relativePath.toString());
+                        if (file.toFile().canExecute()) {
+                            entry.setMode(entry.getMode() | 0755);
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:115]: Are you sure that using `0755` like this is correct?

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        } catch (IOException e) {
+            throw DekorateException.launderThrowable(e);
+        }
+    }
+
+    public static File packageFile(Path root, Path... additional) {
+        return packageFile(root, null, additional);
+    }
+
+    public static File packageFile(Path root, String base, Path... additional) {
```
Comment (geoand) [inline extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:86]: This method seems to have plenty of similarities with `packageFile`. Can the common code be extracted?

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java (hunk 4 above)
Comment (iocanel) [reply extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:26]: This is copied from dekorate and modified so that it can support custom entry prefix paths.
This should be fixed upstream. See: https://github.com/dekorateio/dekorate/issues/621

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java (hunk 5 above)
Comment (iocanel) [reply extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:148]: Same as above.

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java (hunk 7 above)
Comment (iocanel) [reply extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/PackageUtil.java:115]: Yeah

File: extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java (hunk 2 above)
Comment (iocanel) [reply extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java:103]: This is already in the quarkus container image util module.

**Comment by geoand** [comment]
This makes sense to me, but it would be really great if @tqvarnst could give it a spin and make sure it matches his expectations.

File: extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/OpenshiftUtils.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    /**
+     * Merges {@link OpenshiftConfig} with {@link S2iConfig} prioritizing in the former.
+     * 
+     * @param openshiftConfig the Openshift config
+     * @param s2iConfig the s2i config
+     * @param an instance of {@link OpenshiftConfig} with the merged configuration.
+     */
+    public static OpenshiftConfig merge(OpenshiftConfig openshiftConfig, S2iConfig s2iConfig) {
```
Comment (geoand) [inline extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/OpenshiftUtils.java:89]: This name is too vague IMHO. I would prefer if it were named `mergeConfig`

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java [hunk 10]
```diff
@@ -378,6 +383,16 @@ public void build(ApplicationInfoBuildItem applicationInfo,
                     kubernetesHealthLivenessPath,
                     kubernetesHealthReadinessPath);
 
+            decorators.stream().filter(d -> d.matches(Decorator.class)).forEach(i -> {
```
Comment (geoand) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java:386]: What's the rationale of this change? I think we at least need a comment about it

**Review by geoand — COMMENTED** [review]
Great work! I think this is a great approach.

I only added some minor comments

File: extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/BuildStrategy.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+
+package io.quarkus.container.image.openshift.deployment;
+
+public enum BuildStrategy {
+
+    binary,
```
[the remark this answers is not in the document]
Comment (iocanel) [reply extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/BuildStrategy.java:6]: The are lowercase, as they read better when passed to `application.properties`.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java (hunk 10 above)
Comment (iocanel) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java:386]: As I didn't felt comfortable is exposing `Decorate` specific classes in the kubernetes spi. Decorators are passed to build items as Objects. So, at some point we need to check if the build item is indeed holding an instance of the `Decorator.class`. 

Definitely not the most elegant solution, but I feel it's somehow better than dragging decorate to the kubernetes spi.
I am open about it though.

Comment (geoand) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java:386]: Understood. I'm fine with it, but let's please add some relevant comments to the source

**Comment by iocanel** [comment]
@geoand: feedback applied, ci issues resolved!

**Review by geoand — APPROVED** [review]
Excellent work!

@tqvarnst @maxandersen I'm going to merge this and you can give it a spin when you get a chance

**Comment by geoand** [comment]
@iocanel when time comes to release 1.9, we'll need a small piece of text to explain this
